proposals/p3646.md

```diff
@@ -0,0 +1,284 @@
+# Tuples and tuple indexing
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/3646)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Abstract](#abstract)
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+    -   [Lexing](#lexing)
+    -   [Indexes as names](#indexes-as-names)
+    -   [Precedence](#precedence)
+    -   [Expression operand](#expression-operand)
+    -   [Bounds](#bounds)
+    -   [Tuple slicing](#tuple-slicing)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+    -   [Alternative lexing rule](#alternative-lexing-rule)
+    -   [Decimal indexing restriction](#decimal-indexing-restriction)
+    -   [Square bracket notation](#square-bracket-notation)
+    -   [Negative indexing from the end of the tuple](#negative-indexing-from-the-end-of-the-tuple)
+    -   [Trailing commas](#trailing-commas)
+
+<!-- tocstop -->
+
+## Abstract
+
+Add support for extracting elements of a tuple by their numerical index.
+
+Also formally add the well-established basic syntactic and semantic rules for
+tuples, for which we have had leads issues but no proposal, into the design.
+
+## Problem
+
+Currently, the only way to access the elements of a tuple is through pattern
+matching. While this handles many cases well, it is sometimes desirable to
+access an element of a tuple more succinctly, especially in cases where only a
+single element's value is needed.
+
+## Background
+
+In Python, tuple indexing is performed using square brackets:
+
+```python
+tup = (1, 2, 3)
+# Prints 2.
+print(tup[1])
+```
+
+In C++, `std::pair` is indexed using `.first` and `.second`, and `std::tuple` is
+indexed using `std::get<I>`.
+
+In Rust, a tuple is indexed using `.N`, where `N` is a decimal integer literal.
```

Comment:
```suggestion
In both Rust and Swift, a tuple is indexed using `.N`, where `N` is a decimal zero-based integer literal.
```

I think this is true. Is it worth including links? This is a bit minimally documented in Swift, but I found it here:
https://docs.swift.org/swift-book/documentation/the-swift-programming-language/expressions#Explicit-Member-Expression:~:text=The%20members%20of%20a%20tuple%20are%20implicitly%20named%20using%20integers%20in%20the%20order%20they%20appear%2C%20starting%20from%20zero.%20For%20example%3A

Reply:
Interesting. This does seem to be true in Swift, but there's a quirk: Swift does not allow digit separators, nor non-decimal field indexes, but it *does* allow leading zeroes: `(10, 20, 30).01` is permitted and treated as `(10, 20, 30).1`.

I'm not sure if that's intentional or a bug :-)